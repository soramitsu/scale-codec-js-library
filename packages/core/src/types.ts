export type DecodeResult<T> = [value: T, decodedBytes: number];

export type Decode<T> = (bytes: Uint8Array) => DecodeResult<T>;

export type Encode<T> = (value: T) => Uint8Array;

export interface Codec<T> {
    encode: Encode<T>;
    decode: Decode<T>;
}
