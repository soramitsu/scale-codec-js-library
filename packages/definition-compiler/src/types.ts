export interface RenderNamespaceDefinitionParams {
  /**
   * Runtime library with STD codecs + reexports from the core library.
   *
   * Defaults to `@scale-codec/definition-runtime`
   */
  runtimeLib?: string

  /**
   * Types available in the runtime library.
   */
  runtimeTypes?: Set<string>

  /**
   * Single tuples are always an arrays with a single element. It is possible to make final code cleaner
   * (and a bit performant) if render such tuples just as aliases for the inner element. It is optional feature.
   */
  rollupSingleTuplesIntoAliases?: boolean

  /**
   * @default 'Void'
   */
  typeForVoidAliasing?: string

  /**
   * Enable sorting of types to minimize amount of allocated dynamic references between them
   *
   * @default false
   * @experimental
   */
  optimizeDyns?: boolean
}

/**
 * The main map with all the definitions
 *
 * @remarks
 * Name of type should be **a valid JavaScript identifier name**, because it will be compiled to an identifier.
 *
 * @example
 * ```ts
 * const definition: NamespaceDefinition = {
 *   Person: {
 *     t: 'struct',
 *     fields: [
 *       { name: 'name', ref: 'str' },
 *       { name: 'age', ref: 'u8' }
 *     ]
 *   },
 *   VecPerson: { t: 'vec', item: 'Person' }
 * }
 * ```
 */
export type NamespaceDefinition = Record<string, TypeDef>

export type TypeRef = string

/**
 * Just an alias to the inner type
 */
export type DefAlias = {
  ref: TypeRef
}

/**
 * Fixed-length array
 */
export type DefArray = {
  /**
   * Inner type name
   */
  item: TypeRef
  len: number
}

/**
 * It's like {@link DefArray} but for bytes (u8). Use it for bytes for better performance.
 */
export type DefBytesArray = {
  len: number
}

/**
 * `Vec<T>` definition
 */
export type DefVec = {
  /**
   * Inner vec type name
   */
  item: TypeRef
}

/**
 * Tuple definition
 */
export type DefTuple = {
  /**
   * Array of inner types
   */
  items: TypeRef[]
}

/**
 * Structure definition
 */
export type DefStruct = {
  /**
   * @remarks
   * **note**: order of fields matters!
   */
  fields: DefStructField[]
}

export type DefStructField = {
  /**
   * Name of the struct field
   */
  name: string
  /**
   * Reference to the type
   */
  ref: TypeRef
}

/**
 * Map definition (e.g. `HashMap`, `BTreeMap`)
 */
export type DefMap = {
  key: TypeRef
  value: TypeRef
}

/**
 * Set definition (e.g. `HashSet`, `BTreeSet`)
 */
export type DefSet = {
  entry: TypeRef
}

/**
 * Enum definition
 */
export type DefEnum = {
  /**
   * @remarks
   * Order of variants doesn't matter, but variants should not contain collisions between their names
   * and discriminants
   */
  variants: DefEnumVariant[]
}

export type DefEnumVariant = {
  name: string
  discriminant: number
  /**
   * No ref/null ref means that this variant is empty
   */
  ref?: TypeRef | null
}

/**
 * Option enum definition
 */
export type DefOption = {
  some: TypeRef
}

/**
 * Result enum definition
 */
export type DefResult = {
  ok: TypeRef
  err: TypeRef
}

/**
 * External type definition - import codec from external module
 *
 * @remarks
 * Provides a possibility to define external types, e.g. to use some complex structure from another compiled namespace
 * OR to define your own custom low-level codec for type that is not included into the SCALE codec spec by default.
 *
 * Note that the import should be of `FragmentBuilder` type.
 */
export type DefImport = {
  /**
   * Where to import from, path
   *
   * @example
   * ```ts
   * import { ... } from '<here is the module name>'
   * ```
   */
  module: string
  /**
   * Name of the type inside of the module. If this field is omitted, the own type name will be used
   *
   * @todo *define custom name for each import?*
   */
  nameInModule?: string
}

type WithTMark<T, M extends string> = T & {
  t: M
}

export type TypeDef =
  | WithTMark<DefAlias, 'alias'>
  | WithTMark<DefArray, 'array'>
  | WithTMark<DefBytesArray, 'bytes-array'>
  | WithTMark<DefVec, 'vec'>
  | WithTMark<DefTuple, 'tuple'>
  | WithTMark<DefStruct, 'struct'>
  | WithTMark<DefMap, 'map'>
  | WithTMark<DefSet, 'set'>
  | WithTMark<DefEnum, 'enum'>
  | WithTMark<DefOption, 'option'>
  | WithTMark<DefResult, 'result'>
  | WithTMark<DefImport, 'import'>
