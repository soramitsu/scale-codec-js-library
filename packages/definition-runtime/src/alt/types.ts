import {
    Decode,
    Encode,
    Walker,
    encodeFactory,
    WalkerImpl,
    createArrayEncoder,
    createArrayDecoder,
    BigIntTypes,
    createBigIntEncoder,
    createBigIntDecoder,
} from '@scale-codec/core'

import { trackDecode } from '../tracking'

// interface Codec<T> {
//     encode: Encode<T>
//     decode: Decode<T>
//     fromBuffer: (src: ArrayBufferView) => T
//     toBuffer: (val: T) => Uint8Array
// }

class Codec<T> {
    public encode: Encode<T>
    public decode: Decode<T>

    public constructor(name: string, encode: Encode<T>, decode: Decode<T>) {
        this.encode = makeEncodeWithSkipping(encode)
        this.decode = (walker) => trackDecode(name, walker, decode)
    }

    public fromBuffer(src: ArrayBufferView): T {
        return WalkerImpl.decode(src, this.decode)
    }

    public toBuffer(val: T): Uint8Array {
        return WalkerImpl.encode(val, this.encode)
    }
}

const MARK_SKIP_ENCODING = Symbol('SkipEncoding')

interface MarkedToSkipEncoding {
    [MARK_SKIP_ENCODING]: Uint8Array
}

type CodecValue<T extends Codec<any>> = T extends Codec<infer V> ? V : never

function isMarkedToSkip(value: unknown): value is MarkedToSkipEncoding {
    return !!(value && (value as MarkedToSkipEncoding)[MARK_SKIP_ENCODING])
}

function trySkipEncoding(value: unknown, walker: Walker): boolean {
    if (isMarkedToSkip(value)) {
        walker.u8.set(value[MARK_SKIP_ENCODING], walker.idx)
        walker.idx += value[MARK_SKIP_ENCODING].byteLength
        return true
    }
    return false
}

function trySkipEncodingSizeHint(value: unknown): number | undefined {
    if (isMarkedToSkip(value)) {
        return value[MARK_SKIP_ENCODING].byteLength
    }
    return undefined
}

function makeEncodeWithSkipping<T>(encode: Encode<T>): Encode<T | MarkedToSkipEncoding> {
    return encodeFactory(
        (val, walker) => {
            if (trySkipEncoding(val, walker)) return
            encode(val as T, walker)
        },
        (val) => trySkipEncodingSizeHint(val) ?? encode.sizeHint(val as T),
    )
}

function createArrayCodec<T>(name: string, itemCodec: Codec<T>, len: number): Codec<T[]> {
    return new Codec(name, createArrayEncoder(itemCodec.encode, len), createArrayDecoder(itemCodec.decode, len))
}

function createBigIntCodec(ty: BigIntTypes): Codec<bigint> {
    return new Codec(ty, createBigIntEncoder(ty), createBigIntDecoder(ty))
}

const U64: Codec<bigint> = createBigIntCodec('u64')

export const Array_u64_l32: Codec<bigint[]> = createArrayCodec('Array_u64_l32', U64, 32)
