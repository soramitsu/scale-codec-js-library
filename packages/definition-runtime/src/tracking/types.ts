import { Decode, Walker } from '@scale-codec/core'

/**
 * This symbol could be used by tracker to inspect value
 */
export const TrackValueInspect = Symbol('TrackValueInspect')

/**
 * Implement this for the object if you want to handle its inspection in the console
 */
export interface TrackValueInspectable {
  [TrackValueInspect]: () => any
}

export type TrackDecodeFn = <T>(loc: string, walker: Walker, decode: Decode<T>) => T

export type RefineDecodeLocFn = <T>(loc: string, headlessDecode: () => T) => T

/**
 * Instance that defines method to track decoding flow
 */
export interface CodecTracker {
  /**
   * Used to track decode step: location, input, and result
   */
  decode?: TrackDecodeFn
  /**
   * Used to clarify decode location without tracking of the input and the output
   */
  refineDecodeLoc?: RefineDecodeLocFn
}
