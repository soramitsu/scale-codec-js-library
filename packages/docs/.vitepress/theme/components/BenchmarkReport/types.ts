export interface ReportData {
    name: string
    date: string
    version: string | null
    results: Result[]
    fastest: ReportCaseScore
    slowest: ReportCaseScore
}

export interface Result {
    name: string
    ops: number
    margin: number
    percentSlower: number
}

export interface ReportCaseScore {
    name: string
    index: number
}
