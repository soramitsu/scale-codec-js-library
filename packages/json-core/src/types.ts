export type JsonValue = null | string | boolean | number | { [K in string]: JsonValue } | JsonValue[];

export type Serialize<T> = (value: T) => JsonValue;

export type Deserialize<T> = (raw: JsonValue) => T;

export interface Serde<T> {
    ser: Serialize<T>;
    de: Deserialize<T>;
}
