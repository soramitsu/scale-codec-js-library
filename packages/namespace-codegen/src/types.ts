export type NamespaceCodegenDefinition = Record<string, TypeDef>;

export type ArrayDef = {
    item: string;
    len: number;
};

export type VecDef = {
    item: string;
};

export type TupleDef = {
    items: string[];
};

export type StructDef = {
    fields: {
        name: string;
        ref: string;
    }[];
};

export type MapDef = {
    key: string;
    value: string;
};

export type EnumDef = {
    variants: {
        name: string;
        discriminant: number;
        ref: string | null;
    }[];
};

export type OptionDef = {
    some: string;
};

export type ResultDef = {
    ok: string;
    err: string;
};

export type WithTMark<T, M extends string> = T & {
    t: M;
};

export type TypeDef =
    | string
    | WithTMark<ArrayDef, 'array'>
    | WithTMark<VecDef, 'vec'>
    | WithTMark<TupleDef, 'tuple'>
    | WithTMark<StructDef, 'struct'>
    | WithTMark<MapDef, 'map'>
    | WithTMark<EnumDef, 'enum'>
    | WithTMark<OptionDef, 'option'>
    | WithTMark<ResultDef, 'result'>;
