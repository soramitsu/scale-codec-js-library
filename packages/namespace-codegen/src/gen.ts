import { NamespaceCodegenDefinition } from './types';
import { DefEnum, typeDefToEnum } from './def-enum';
import { assert } from '@scale-codec/util';
import { StdCodecs } from '@scale-codec/namespace';
import { createImportsCounter, getImportsCounterState, ImportsCounter } from './imports-counter';

export interface GenerateOptions {
    namespaceTypeName: string;
    namespaceValueName: string;
    importLib: string;
    structPropsCamelCase?: boolean;
}

type DefMap = Map<string, DefEnum>;

const STD_KEYS = new Set<string>(Object.keys(StdCodecs));

const DEFAULT_STD_IMPORTS = new Set(['StdTypes', 'StdCodecs', 'defNamespace']);

/**
 * TODO batch all validation errors and throw once?
 */
function validate(defmap: DefMap) {
    function assertReference(ref: string, from: string) {
        assert(defmap.has(ref) || STD_KEYS.has(ref), `Undefined reference "${ref}" from "${from}"`);
    }

    for (const [typeName, def] of defmap) {
        assert(!STD_KEYS.has(typeName), `Type name "${typeName}" is forbidden - reserved for STD`);

        def.match({
            Alias(target) {
                assertReference(target, `${typeName} (alias)`);
            },
            Vec({ item }) {
                assertReference(item, `${typeName} (vec)`);
            },
            Array({ item }) {
                assertReference(item, `${typeName} (array)`);
            },
            Tuple({ items }) {
                items.forEach((ref, i) => {
                    assertReference(ref, `${typeName}[${i}] (tuple)`);
                });
            },
            Struct({ fields }) {
                const namesEncountered = new Set<string>();

                fields.forEach(({ name, ref }, i) => {
                    assert(!namesEncountered.has(name), `Duplicated field "${name}" in struct "${typeName}", pos ${i}`);
                    assertReference(ref, `${typeName}.${name} (struct)`);
                });
            },
            Map({ key, value }) {
                assertReference(key, `${typeName} (map key)`);
                assertReference(value, `${typeName} (map value)`);
            },
            Enum({ variants }) {
                // check no discriminant collisions
                // check no name collisions
                // check references

                const discriminantsEncountered = new Set<number>();
                const variantsEncountered = new Set<string>();

                variants.forEach(({ name: variant, discriminant, ref }, i) => {
                    assert(
                        !Number.isNaN(discriminant) && discriminant >= 0 && Number.isInteger(discriminant),
                        `Discriminant should be valid non-negative integer, found: ${discriminant} (enum ${typeName}::${variant})`,
                    );
                    assert(
                        !variantsEncountered.has(variant),
                        `Duplicated variant "${variant}" in enum "${typeName}", pos ${i}`,
                    );
                    assert(
                        !discriminantsEncountered.has(discriminant),
                        `Duplicated discriminant ${discriminant} in enum, "${typeName}::${variant}", pos ${i}`,
                    );
                    ref && assertReference(ref, `${typeName}::${variant} (enum)`);
                });
            },
            EnumOption({ some }) {
                assertReference(some, `${typeName} (option)`);
                assertReference(some, `${typeName} (option)`);
            },
            EnumResult({ ok, err }) {
                assertReference(ok, `${typeName} (result ok)`);
                assertReference(err, `${typeName} (result err)`);
            },
        });
    }
}

function genImports(set: Set<string>, importFrom: string) {
    const items = [...set];
    items.sort();
    return `import { ${items.join(', ')} } from '${importFrom}';`;
}

function genNamespaceDeclaration(
    defMap: DefMap,
    imports: ImportsCounter,
    opts: {
        namespaceName: string;
    },
): string {
    const NAMESPACE_NAME = opts.namespaceName;

    function nsItem(ref: string): string {
        return `${NAMESPACE_NAME}['${ref}']`;
    }

    const lines = [`export type ${NAMESPACE_NAME} = StdTypes & {`];

    for (const [key, def] of defMap) {
        const defToType: string = def.match({
            Alias(to) {
                return nsItem(to);
            },
            Vec({ item }) {
                return `${nsItem(item)}[]`;
            },
            Array({ item }) {
                return `${nsItem(item)}[]`;
            },
            Tuple({ items }) {
                const mapped = items.map(nsItem);

                return `[${mapped.join(',')}]`;
            },
            Struct({ fields }) {
                const fieldsAsEntries: string[] = fields.map(
                    // TODO camelCase?
                    ({ name, ref }) => `${name}: ${nsItem(ref)}`,
                );

                return `{\n${fieldsAsEntries.join(';\n')} }`;
            },
            Map({ key, value }) {
                return `Map<${nsItem(key)}, ${nsItem(value)}>`;
            },
            Enum({ variants }) {
                const tsVariants: string[] = variants.map(
                    ({ name, ref }) => `${name}: ${ref && `${imports.Valuable}<${nsItem(ref)}>`}`,
                );

                return `${imports.Enum}<{\n${tsVariants.join(';\n')} }>`;
            },
            EnumOption({ some }) {
                return `${imports.Option}<${nsItem(some)}>`;
            },
            EnumResult({ ok, err }) {
                return `${imports.Result}<${nsItem(ok)}, ${nsItem(err)}>`;
            },
        });

        lines.push(`    "${key}": ${defToType};`);
    }

    lines.push('}');

    return lines.join('\n');
}

function genNamespaceValue(
    defmap: DefMap,
    imports: ImportsCounter,
    opts: {
        namespaceType: string;
        namespaceValue: string;
    },
): string {
    // const additionalStdImports = new Set<string>();
    // const helpersProxy = new Proxy(StdDefHelpers, {
    //     get(target, key) {
    //         if (!(key in target)) throw new Error('wtf?');
    //         additionalStdImports.add(key as string);
    //         return target[key as keyof typeof StdDefHelpers];
    //     },
    // }) as typeof StdDefHelpers;

    const lines = [`export const ${opts.namespaceValue} = defNamespace<${opts.namespaceType}>({`, '...StdCodecs,'];

    for (const [key, def] of defmap) {
        const defToType: string = def.match({
            Alias(to) {
                return `${imports.defAlias}('${to}')`;
            },
            Vec({ item }) {
                return `${imports.defVec}('${item}')`;
            },
            Array({ item, len }) {
                return `${imports.defArray}('${item}', ${len})`;
            },
            Tuple({ items }) {
                return `${imports.defTuple}(${items.map((x) => `'${x}'`).join(', ')})`;
            },
            Struct({ fields }) {
                const fieldsAsTuples = fields.map(({ name, ref }) => `['${name}', '${ref}']`).join(',\n');

                return `${imports.defStruct}([\n${fieldsAsTuples}])`;
            },
            Map({ key, value }) {
                return `${imports.defMap}('${key}', '${value}')`;
            },
            Enum({ variants }) {
                const variantsWithDiscriminants = variants
                    .map(({ name, discriminant }) => `${name}: { discriminant: ${discriminant} }`)
                    .join(',');
                const schema = `new ${imports.EnumSchema}({\n${variantsWithDiscriminants} })`;

                const codecsOfValuable = variants
                    .filter((x) => !!x.ref)
                    .map(({ name, ref }) => `${name}: '${ref}'`)
                    .join(',\n');
                const codecs = `{\n${codecsOfValuable} }`;

                return `${imports.defEnum}(${schema}, ${codecs})`;
            },
            EnumOption({ some }) {
                return `${imports.defOption}('${some}')`;
            },
            EnumResult({ ok, err }) {
                return `${imports.defResult}('${ok}', '${err}')`;
            },
        });

        lines.push(`    "${key}": ${defToType},`);
    }

    lines.push('});');

    return lines.join('\n');
}

export function generate(definition: NamespaceCodegenDefinition, opts: GenerateOptions): string {
    // collect all known keys and map definitions into enums!
    // then generate main namespace type, map with types. chech on this stage that all
    // references are valid and report friendly errors
    // then generate namespace actual declaration

    const definitionMap: DefMap = new Map(Object.entries(definition).map(([key, def]) => [key, typeDefToEnum(def)]));
    const importsCounter = createImportsCounter();

    validate(definitionMap);

    const typeDec = genNamespaceDeclaration(definitionMap, importsCounter, {
        namespaceName: opts.namespaceTypeName,
    });

    const valDef = genNamespaceValue(definitionMap, importsCounter, {
        namespaceValue: opts.namespaceValueName,
        namespaceType: opts.namespaceTypeName,
    });

    const { used: additionalStdImports } = getImportsCounterState(importsCounter);

    const imports = genImports(new Set([...additionalStdImports, ...DEFAULT_STD_IMPORTS]), opts.importLib);

    return `${imports}\n\n${typeDec}\n\n${valDef};`;
}
