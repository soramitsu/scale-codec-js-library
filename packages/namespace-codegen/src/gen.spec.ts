import prettier from 'prettier';
import prettierConfig from '../../../.prettierrc.js';
import { NamespaceCodegenDefinition } from './types';
import { generate, GenerateOptions } from './gen';

function format(tsCode: string): string {
    return prettier.format(tsCode, {
        ...(prettierConfig as any),
        parser: 'typescript',
    });
}

function shouldMatchCode(def: NamespaceCodegenDefinition, opts: GenerateOptions, expectedLike: string) {
    const expected = format(expectedLike);
    const actual = format(generate(def, opts));

    expect(actual).toEqual(expected);
}

function shouldMatchSnapshot(def: NamespaceCodegenDefinition, opts: GenerateOptions) {
    expect(format(generate(def, opts))).toMatchSnapshot();
}

function sortImports(...imports: string[]): string {
    imports.sort();
    return imports.join(', ');
}

describe('generate()', () => {
    test('empty definition', () => {
        shouldMatchCode(
            {},
            {
                namespaceTypeName: 'CustomNamespace',
                namespaceValueName: 'ns',
                importLib: 'scale-codec',
            },
            `
            import { ${sortImports('StdTypes', 'StdCodecs', 'defNamespace')} } from 'scale-codec';

            export type CustomNamespace = StdTypes & {};

            export const ns = defNamespace<CustomNamespace>({
                ...StdCodecs
            })
            `,
        );
    });

    test('definition with alias to std', () => {
        shouldMatchCode(
            { String: 'str' },
            { namespaceTypeName: 'WithAlias', namespaceValueName: 'ns', importLib: 'scale-codec' },
            `
            import { ${sortImports('StdTypes', 'StdCodecs', 'defNamespace', 'defAlias')} } from 'scale-codec';

            export type WithAlias = StdTypes & {
                String: WithAlias['str']
            };

            export const ns = defNamespace<WithAlias>({
                ...StdCodecs, String: defAlias('str') })
            `,
        );
    });

    test('definition with struct', () => {
        shouldMatchSnapshot(
            {
                Id: {
                    t: 'struct',
                    fields: [
                        {
                            name: 'id',
                            ref: 'str',
                        },
                    ],
                },
            },
            {
                namespaceTypeName: 'WithId',
                namespaceValueName: 'withId',
                importLib: 'my-scale',
            },
        );
    });

    test('complex definition', () => {
        shouldMatchSnapshot(
            {
                String: 'str',
                '(u64, u32)': {
                    t: 'tuple',
                    items: ['u64', 'u32'],
                },
                'BTreeMap<String, bool>': {
                    t: 'map',
                    key: 'String',
                    value: 'bool',
                },
                'Option<(u64, u32)>': {
                    t: 'option',
                    some: '(u64, u32)',
                },
                'Result<String, String>': {
                    t: 'result',
                    ok: 'String',
                    err: 'String',
                },
                'Result<(), str>': {
                    t: 'result',
                    ok: '()',
                    err: 'str',
                },
                IpAddr: {
                    t: 'enum',
                    variants: [
                        {
                            name: 'V4',
                            discriminant: 0,
                            ref: 'u8',
                        },
                        {
                            name: 'V6',
                            discriminant: 1,
                            ref: 'u16',
                        },
                        {
                            name: 'Invalid',
                            discriminant: 3,
                            ref: null,
                        },
                    ],
                },
            },
            {
                namespaceTypeName: 'Complex',
                namespaceValueName: 'complex',
                importLib: '@scale-codec/namespace',
            },
        );
    });

    test('error if entry duplicated std type', () => {
        expect(() =>
            generate(
                {
                    str: 'u32',
                },
                { namespaceTypeName: 'test', namespaceValueName: 'test', importLib: 'test' },
            ),
        ).toThrow();
    });

    test('error if duplicated enum discriminants', () => {
        expect(() => {
            generate(
                {
                    SomeEnum: {
                        t: 'enum',
                        variants: [
                            {
                                name: 'one',
                                discriminant: 0,
                                ref: null,
                            },
                            {
                                name: 'two',
                                discriminant: 0,
                                ref: null,
                            },
                        ],
                    },
                },
                { namespaceTypeName: 'test', namespaceValueName: 'test', importLib: 'test' },
            );
        }).toThrow();
    });

    test.todo('other validation errors');

    test('struct props camelCased if related option is provided', () => {
        shouldMatchSnapshot(
            {
                Example: {
                    t: 'struct',
                    fields: [
                        {
                            name: 'foo_bar_baz',
                            ref: 'u64',
                        },
                        {
                            name: 'another_snake',
                            ref: 'u32',
                        },
                    ],
                },
            },
            {
                namespaceTypeName: 'test',
                namespaceValueName: 'test',
                importLib: 'test',

                structPropsCamelCase: true,
            },
        );
    });

    test('using set', () => {
        shouldMatchSnapshot(
            {
                'BTreeSet<str>': {
                    t: 'set',
                    entry: 'str',
                },
            },
            {
                namespaceTypeName: 'test',
                namespaceValueName: 'test',
                importLib: 'test',
            },
        );
    });

    test('array of bytes', () => {
        shouldMatchSnapshot(
            {
                '[u8; 512]': {
                    t: 'bytes-array',
                    len: 512,
                },
            },
            {
                namespaceTypeName: 'TYPE',
                namespaceValueName: 'VALUE',
                importLib: 'test',
            },
        );
    });
});
