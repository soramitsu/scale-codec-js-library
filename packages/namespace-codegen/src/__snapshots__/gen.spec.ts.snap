// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate() array of bytes 1`] = `
"import { StdCodecs, StdTypes, defBytesArray, defNamespace } from 'test';

export type TYPE = StdTypes & {
    '[u8; 512]': Uint8Array;
};

export const VALUE = defNamespace<TYPE>({
    ...StdCodecs,
    '[u8; 512]': defBytesArray(512),
});
"
`;

exports[`generate() complex definition 1`] = `
"import {
    Enum,
    EnumSchema,
    Option,
    Result,
    StdCodecs,
    StdTypes,
    Valuable,
    defAlias,
    defEnum,
    defMap,
    defNamespace,
    defOption,
    defResult,
    defTuple,
} from '@scale-codec/namespace';

export type Complex = StdTypes & {
    String: Complex['str'];
    '(u64, u32)': [Complex['u64'], Complex['u32']];
    'BTreeMap<String, bool>': Map<Complex['String'], Complex['bool']>;
    'Option<(u64, u32)>': Option<Complex['(u64, u32)']>;
    'Result<String, String>': Result<Complex['String'], Complex['String']>;
    'Result<(), str>': Result<Complex['()'], Complex['str']>;
    IpAddr: Enum<{
        V4: Valuable<Complex['u8']>;
        V6: Valuable<Complex['u16']>;
        Invalid: null;
    }>;
};

export const complex = defNamespace<Complex>({
    ...StdCodecs,
    String: defAlias('str'),
    '(u64, u32)': defTuple(['u64', 'u32']),
    'BTreeMap<String, bool>': defMap('String', 'bool'),
    'Option<(u64, u32)>': defOption('(u64, u32)'),
    'Result<String, String>': defResult('String', 'String'),
    'Result<(), str>': defResult('()', 'str'),
    IpAddr: defEnum(
        new EnumSchema({
            V4: { discriminant: 0 },
            V6: { discriminant: 1 },
            Invalid: { discriminant: 3 },
        }),
        {
            V4: 'u8',
            V6: 'u16',
        },
    ),
});
"
`;

exports[`generate() definition with struct 1`] = `
"import { StdCodecs, StdTypes, defNamespace, defStruct } from 'my-scale';

export type WithId = StdTypes & {
    Id: {
        id: WithId['str'];
    };
};

export const withId = defNamespace<WithId>({
    ...StdCodecs,
    Id: defStruct([['id', 'str']]),
});
"
`;

exports[`generate() struct props camelCased if related option is provided 1`] = `
"import { StdCodecs, StdTypes, defNamespace, defStruct } from 'test';

export type test = StdTypes & {
    Example: {
        fooBarBaz: test['u64'];
        anotherSnake: test['u32'];
    };
};

export const test = defNamespace<test>({
    ...StdCodecs,
    Example: defStruct([
        ['fooBarBaz', 'u64'],
        ['anotherSnake', 'u32'],
    ]),
});
"
`;

exports[`generate() using set 1`] = `
"import { StdCodecs, StdTypes, defNamespace, defSet } from 'test';

export type test = StdTypes & {
    'BTreeSet<str>': Set<test['str']>;
};

export const test = defNamespace<test>({
    ...StdCodecs,
    'BTreeSet<str>': defSet('str'),
});
"
`;
